File: exercism/elixir/list-ops/list_ops.exs
defmodule ListOps do
  # Please don't use any external modules (especially List) in your
  # implementation. The point of this exercise is to create these basic functions
  # yourself.
  #
  # Note that `++` is a function from an external module (Kernel, which is
  # automatically imported) and so shouldn't be used either.

  @spec count(list) :: non_neg_integer
  def count(l) do
    trav = fn(ls, trav) ->
      case ls do
        [h | t] -> 1 + trav.(t, trav)
        [] -> 0
      end
    end

    trav.(l, trav)
  end

  @spec reverse(list) :: list
  def reverse(l) do
    trav = fn(ls, acc, trav) ->
      case ls do
        [h | t] -> trav.(t, [h | acc], trav)
        [] -> acc
      end
    end

    trav.(l, [], trav)
  end

  @spec map(list, (any -> any)) :: list
  def map(l, f) do
    trav = fn(ls, trav) ->
      case ls do
        [h | t] -> [f.(h) | trav.(t,trav)]
        [] -> []
      end
    end

    trav.(l, trav)
  end

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter(l, f) do
    trav = fn(ls, trav) ->
      case ls do
        [h | t] ->
          case f.(h)  do
            true -> [h | trav.(t,trav)]
            _ -> trav.(t,trav)
          end
        [] -> []
      end
    end

    trav.(l, trav)
  end

  @type acc :: any
  @spec reduce(list, acc, ((any, acc) -> acc)) :: acc
  def reduce(l, acc, f) do
    trav = fn(ls, acc, trav) ->
      case ls do
        [h | t] -> trav.(t, f.(h, acc), trav)
        [] -> acc
      end
    end

    trav.(l, acc, trav)
  end

  @spec append(list, list) :: list
  def append(a, b) do
    trav = fn(ls, trav) ->
      case ls do
        [h | t] -> [h | trav.(t, trav)]
        [] -> b
      end
    end

    trav.(a, trav)
  end

  @spec concat([[any]]) :: [any]
  def concat(ll) do
    trav = fn(ls, trav) ->
      lss = cond do
        is_map(ls) -> List.to_list(ls)
        true -> ls
      end

      case lss do
        [h | t] ->
          case is_list(h) do
            true -> append(h, trav.(t, trav))
            false -> [h | trav.(t, trav)]
          end
        [] -> []
      end
    end

    trav.(ll, trav)
  end
end

File: exercism/haskell/list-ops/src/ListOps.hs
module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ acc [] = acc
foldl' f acc (x:xs) = foldl' f nextVal xs
  where nextVal = f acc x

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ acc [] = acc
foldr f acc (x:xs) = f x nextVal
  where nextVal = foldr f acc xs

length :: [a] -> Int
length = foldl' (\a _ -> 1 + a) 0

reverse :: [a] -> [a]
reverse = foldl (\acc a -> a : acc) []

map :: (a -> b) -> [a] -> [b]
map f = foldr (\x acc -> f x : acc) []

filter :: (a -> Bool) -> [a] -> [a]
filter f = foldl' (\acc x -> next acc x) []
  where next acc x = case f x of
          True ->  x : acc
          _ ->  acc

(++) :: [a] -> [a] -> [a]
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldr (++) []
